core/src/main/java/com/sequenceiq/cloudbreak/converter/stack/cluster/gateway/GatewayJsonToGatewayConverter.java

```diff
@@ -1,17 +1,83 @@
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.GatewayTopology;
 
 @Component
 public class GatewayJsonToGatewayConverter extends AbstractConversionServiceAwareConverter<GatewayJson, Gateway> {
 
     @Override
     public Gateway convert(GatewayJson source) {
+        Gateway gateway = new Gateway();
+        if (source.getGatewayType() != null) {
+            gateway.setGatewayType(source.getGatewayType());
+        }
+        if (source.getSsoType() != null) {
+            gateway.setSsoType(source.getSsoType());
+        } else {
+            gateway.setSsoType(SSOType.NONE);
+        }
+        gateway.setTokenCert(source.getTokenCert());
+        return convertTopologies(source, gateway);
+    }
+
+    private Gateway convertTopologies(GatewayJson gatewayJson, Gateway gateway) {
+        if (isLegacyGatewayRequest(gatewayJson)) {
+            if (!gatewayJson.isEnableGateway()) {
+                return null;
+            }
+            return convertLegacyTopology(gatewayJson, gateway);
+        } else {
+            return convertTopologyList(gatewayJson, gateway);
+        }
+    }
+
+    private boolean isLegacyGatewayRequest(GatewayJson source) {
+        return StringUtils.isNotBlank(source.getTopologyName()) || !CollectionUtils.isEmpty(source.getExposedServices());
+    }
+
+    private Gateway convertLegacyTopology(GatewayJson gatewayJson, Gateway gateway) {
+        String topologyName = StringUtils.isEmpty(gatewayJson.getTopologyName())
+                ? "services"
+                : gatewayJson.getTopologyName();
+        GatewayTopology convertedTopology = CollectionUtils.isEmpty(gatewayJson.getExposedServices())
+                ? new GatewayTopology()
+                : doLegacyConversion(gatewayJson, topologyName);
+        convertedTopology.setTopologyName(topologyName);
+        convertedTopology.setGateway(gateway);
+        gateway.setTopologies(Collections.singleton(convertedTopology));
+        return gateway;
+    }
+
+    private GatewayTopology doLegacyConversion(GatewayJson gatewayJson, String topologyName) {
+        GatewayTopologyJson legacyTopology = new GatewayTopologyJson();
+        legacyTopology.setTopologyName(topologyName);
+        legacyTopology.setExposedServices(gatewayJson.getExposedServices());
+        return getConversionService().convert(legacyTopology, GatewayTopology.class);
+    }
 
-        return null;
+    private Gateway convertTopologyList(GatewayJson gatewayJson, Gateway gateway) {
+        List<GatewayTopologyJson> topologies = gatewayJson.getTopologies();
+        if (!CollectionUtils.isEmpty(topologies)) {
+            Set<GatewayTopology> gatewayTopologies = topologies.stream()
+                    .map(g -> getConversionService().convert(g, GatewayTopology.class))
+                    .collect(Collectors.toSet());
+            gateway.setTopologies(gatewayTopologies);
+            gatewayTopologies.forEach(g -> g.setGateway(gateway));
```

Comment:
Sets all of them but only last one is used later.

Reply:
what do you mean? all the topologies matter